gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/MessageConsumerImpl.java

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import static com.google.cloud.pubsub.spi.v1.SubscriberApi.formatSubscriptionName;
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+import com.google.cloud.GrpcServiceOptions.ExecutorFactory;
+import com.google.cloud.pubsub.PubSub.MessageConsumer;
+import com.google.cloud.pubsub.PubSub.MessageProcessor;
+import com.google.cloud.pubsub.spi.PubSubRpc;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+
+import io.grpc.internal.SharedResourceHolder;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Default implementation for a message consumer.
+ */
+final class MessageConsumerImpl implements MessageConsumer {
+
+  private static final int MAX_QUEUED_CALLBACKS = 100;
+  // shared scheduled executor, used to schedule pulls
+  private static final SharedResourceHolder.Resource<ScheduledExecutorService> TIMER =
+      new SharedResourceHolder.Resource<ScheduledExecutorService>() {
+        @Override
+        public ScheduledExecutorService create() {
+          ScheduledThreadPoolExecutor timer = new ScheduledThreadPoolExecutor(1);
+          timer.setRemoveOnCancelPolicy(true);
+          return timer;
+        }
+
+        @Override
+        public void close(ScheduledExecutorService instance) {
+          instance.shutdown();
+        }
+      };
+
+  private final PubSubOptions pubsubOptions;
+  private final PubSubRpc pubsubRpc;
+  private final PubSub pubsub;
+  private final AckDeadlineRenewer deadlineRenewer;
+  private final String subscription;
+  private final MessageProcessor messageProcessor;
+  private final ScheduledExecutorService timer;
+  private final ExecutorFactory<ExecutorService> executorFactory;
+  private final ExecutorService executor;
+  private final AtomicInteger queuedCallbacks;
+  private final int maxQueuedCallbacks;
+  private final Object futureLock = new Object();
+  private Runnable puller;
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Private field 'puller' could be made final; it is only initialized in the declaration or constructor.](https://www.codacy.com/app/mziccard/gcloud-java/file/2708623361/issues/source?bid=3123109&fileBranchId=3343705#l74)